src/lib/operators/aggregate_hash.cpp

```diff
@@ -215,54 +215,89 @@ void AggregateHash::_aggregate() {
       const auto column_id = _groupby_column_ids.at(group_column_index);
       const auto data_type = input_table->column_data_type(column_id);
 
+      const auto chunk_count = input_table->chunk_count();
+
       resolve_data_type(data_type, [&](auto type) {
         using ColumnDataType = typename decltype(type)::type;
 
-        /*
-        Store unique IDs for equal values in the groupby column (similar to dictionary encoding).
-        The ID 0 is reserved for NULL values. The combined IDs build an AggregateKey for each row.
-        */
-
-        // This time, we have no idea how much space we need, so we take some memory and then rely on the automatic
-        // resizing. The size is quite random, but since single memory allocations do not cost too much, we rather
-        // allocate a bit too much.
-        auto temp_buffer = boost::container::pmr::monotonic_buffer_resource(1'000'000);
-        auto allocator = PolymorphicAllocator<std::pair<const ColumnDataType, AggregateKeyEntry>>{&temp_buffer};
-
-        auto id_map = std::unordered_map<ColumnDataType, AggregateKeyEntry, std::hash<ColumnDataType>, std::equal_to<>,
-                                         decltype(allocator)>(allocator);
-        AggregateKeyEntry id_counter = 1u;
-
-        const auto chunk_count = input_table->chunk_count();
-        for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
-          const auto chunk_in = input_table->get_chunk(chunk_id);
-          if (!chunk_in) continue;
+        if constexpr (std::is_same_v<ColumnDataType, int32_t>) {
+          // For values with a type than AggregateKeyEntry, we can use the value itself as an AggregateKeyEntry. We
+          // cannot do this for types with the same size as AggregateKeyEntry as we need tohave a special NULL value.
+          // By using the value itself, we can save us the effort of building the id_map.
+          for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
+            const auto chunk_in = input_table->get_chunk(chunk_id);
+            const auto base_segment = chunk_in->get_segment(column_id);
+            ChunkOffset chunk_offset{0};
+            segment_iterate<ColumnDataType>(*base_segment, [&](const auto& position) {
+              const auto normalize = [](const int32_t value) {
+                // We need to convert a potentially negative int32_t value into the uint64_t space. We do not care
+                // about preserving the value, just its uniqueness.
+                const auto shifted_value = static_cast<int64_t>(value) - std::numeric_limits<int32_t>::lowest();
```

Comment:
Not sure if "normalize" is the best term here. And is that an implicit "underflow" you are using here?